libs/note-manager/src/note-manager.ts

```diff
@@ -30,39 +36,79 @@ export class NoteManager {
     this.notesMap = new Map();
   }
 
+  /**
+   * Get the resource id of the note
+   * @param note the note to parse and get the resource id
+   * @param isSource if true, get the resource id of the source chain,
+   * otherwise get the resource id of the target chain
+   */
+  static getResourceId(note: Note, isSource?: boolean): ResourceId {
+    const typedChainId = isSource
+      ? note.note.sourceChainId
+      : note.note.targetChainId;
+    const { chainId, chainType } = parseTypedChainId(+typedChainId);
+
+    const contractAddress = isSource
+      ? note.note.sourceIdentifyingData
+      : note.note.targetIdentifyingData;
+
+    return new ResourceId(contractAddress, chainType, chainId);
+  }
+
   static async initAndDecryptNotes(
```

Comment:
A couple improvement considerations for this function:

1. Use `Object.fromEntries` instead of `reduce` to create the `encryptedNotesMap`.

2. Change the `if` statement that checks if `encryptedNotesMap` is truthy to if `(Object.keys(encryptedNotesMap).length > 0)` to make sure that there are actually encrypted notes.

3. Use `Promise.allSettled` instead of `Promise.all` to make sure that all decryption attempts are completed, even if some fail. This can prevent the function from crashing if there is an error with a single note.

4. Add error handling to the decryption process to catch any errors that may occur during decryption.

Reply:
The `Object.fromEntries` does not accept a callback for us to add our own logic to filter the notes which are indexed by `ResourceId`. All other comments are resolved as suggested.